src/Map.js

```diff
@@ -743,16 +743,20 @@ export function deepMerger(existing, value, key) {
     is(existing, value) ? existing : value;
 }
 
-export function deepMergerWith(merger) {
+export function deepMergerWith(merger, keyPath) {
   return (existing, value, key) => {
     if (existing && existing.mergeDeepWith && isIterable(value)) {
-      return existing.mergeDeepWith(merger, value);
+      return mergeDeepWithKeyPath(existing, merger, keyPath.concat(key), value);
     }
-    var nextValue = merger(existing, value, key);
+    var nextValue = merger(existing, value, key, keyPath.concat(key));
     return is(existing, nextValue) ? existing : nextValue;
   };
 }
 
+export function mergeDeepWithKeyPath(existing, merger, keyPath, ...iters) {
+  return existing.mergeWith(deepMergerWith(merger, keyPath), ...iters);
+}
```

Comment:
This avoids the keyPath getting passed directly to mergeDeepWith. It also avoids adding another method to map and list, while still allowing a call to the custom mergeIntoMapWith/ListWith functions by using mergeWith. Not sure if this is the best solution, but I definitely like it more than before.